pyterrier/io.py

```diff
@@ -44,6 +46,23 @@ def find_files(dir):
             files.append(os.path.join(dirpath, name))
     return sorted(files)
 
+
+@contextmanager
+def finialized_open(path, mode):
+    assert mode in ('b', 't') # must supply either binary or text mode
+    # adapted from <https://github.com/allenai/ir_datasets/blob/master/ir_datasets/util/__init__.py#L34>
+    try:
+        with open(f'{path}.tmp', f'x{mode}') as f: # open in exclusive write mode (raises error if already exists)
```

Comment:
I don't think we want it to use autoopen here, right?

Let's say we're downloading `somefile.gz`. If this used autoopen, wouldn't it gzip compress the file on top of the compression already applied?

Reply:
Note: the previous implementation didn't use autoopen either, probably for the same reason: https://github.com/terrier-org/pyterrier/pull/216/commits/72881bda0217203ce2e567679e148d162d636a75#diff-aca08d98f5119d817be6dc6b695ff982dea57aecfc1cc8cbc69f8344d7d15f25L141

Reply:
Then probably it shouldnt be in pt.io if its specific to datasets?

Reply:
It's still pretty general-purpose though. In IRDS, I try to always use it when writing files in case there's an error -- to handle concurrency and to be sure there are not partially-built files laying around if something goes wrong.

Maybe there should also be a `finialized_autoopen`? And/or add an `open_handler=pt.autoopen` parameter and set it to `open` in datasets?